WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift

```diff
@@ -79,3 +79,17 @@ extension Product: ProductFormDataModel {
         productStockStatus
     }
 }
+
+extension ProductVariation: ProductFormDataModel {
```

Comment:
⚠️  One concern with the `ProductFormDataModel` `protocol` approach is that it adds additional properties to `ProductVariation` itself. Considering the [multitude of properties](https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Model/Product/ProductVariation.swift) that `ProductVariation` already has, the lines between "which ones are from the API?" and "which ones are owned by us" get blurry. I can see myself getting confused by these properties in the future. 

For example, in this test, I had thought that `isShippingEnabled` is just a regular property. 😅 

https://github.com/woocommerce/woocommerce-ios/blob/d9325a2751567caff3587dcd484fe580ab8e498c/WooCommerce/WooCommerceTests/ViewRelated/Products/ProductVariation%2BProductFormTests.swift#L34-L37

Also, consider the fact that this is an `extension`. This means that I can now access `isShippingEnabled` from **anywhere** in the WooCommerce package. For example, in here:

https://github.com/woocommerce/woocommerce-ios/blob/d9325a2751567caff3587dcd484fe580ab8e498c/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsDataSource.swift#L392-L401

I can add: 

```swift
        let isShippingEnabled = product?.isShippingEnabled
```

I don't think it even makes sense to access that property in `OrderDetailsDataSource`. But I can. 🤷 I believe this adds confusion for us down the road. 

This is why I'm always wary of extensions. And would try to shy away from them if there's another way. 

For this specific case, one idea that comes to mind is to consider encapsulating the `Product` and `ProductVariation` inside their own respective classes instead — specifically for product editing. And then those classes can be joined by a `protocol`. And it is that `protocol` that we will always interact with in the product editing module. 

Here's a rough example:

```swift
protocol EditableProductEntity {
    var description: String? { get }
    var shortDescription: String? { get }
}

final class EditableProduct: EditableProductEntity {
    private let product: Product

    init(product: Product) {
        self.product = product
    }

    var description: String? {
        product.fullDescription
    }

    var shortDescription: String? {
        product.shortDescription
    }
}

final class EditableProductVariation: EditableProductEntity {
    private let variation: ProductVariation

    init(variation: ProductVariation) {
        self.variation = variation
    }

    var description: String? {
        variation.description
    }

    var shortDescription: String? {
        nil
    }
}
```

It's a bit longer, that's true. But I believe it's sometimes worth it to have longer code in order to maintain a **clean namespace**.

Reply:
yea it might be nice to differentiate the helpers vs. API properties. how would you have the `isShippingEnabled` in the `EditableProductEntity` example above? I'm still trying to understand how the encapsulation helps with the situation 🤔 do you think updating the extension helper variables to functions might help? so that the properties are from the API, and functions are helpers / derived values.

personally, I think it's okay for anyone to access the extension helpers in this case - like if the order details might use `isShippingEnabled` to configure its product related UI, that's totally fine. on the other hand, I could see that we might not want certain helpers to be accessible to other classes. I could also move these extensions to where they are used and mark them private, at the cost of removing the unit tests.

leaving this PR open so that I can update based on the feedback 🙂

Reply:
> I could also move these extensions to where they are used and mark them private, at the cost of removing the unit tests.

This was the first thing that I tried first. It didn't work for me. It was saying that I can't add `private` extensions to conform to `ProductFormDataModel`, which has an `internal` access level. ¯\\\_(ツ)\_/¯

A possible solution would be having the product editing as a separate module (Swift Package Manager?). But I think we're a long way from that. 

> personally, I think it's okay for anyone to access the extension helpers in this case

Yeah, this might be okay for now since the project is not that big yet. I'm just highlighting this now as I have seen how extensions can be abused if we are not vigilant. WordPress is probably a good example of this:

<img src="https://user-images.githubusercontent.com/198826/88865782-07094800-d1c6-11ea-9f01-7472b9afbd73.png" width="540">

You'll notice that some of those extensions are not related to each other and are **tightly bound** to a specific feature in the app. The _extensions_ that they add are not even useful anywhere else. 

> how would you have the isShippingEnabled in the EditableProductEntity example above?

It will be like the `description` and `shortDescription`. 

```swift
    var isShippingEnabled: Bool {
        product.downloadable == false && product.virtual == false
    }
```

As I said, there's going to be a lot of code. And I guess I didn't mention the duplication. 😅 

We can hide the duplication by (surprise!) another `protocol`. This time, a stricter one that only defines properties that are truly defined in both `Product` and `Variation`.

![image](https://user-images.githubusercontent.com/198826/88866805-fa3a2380-d1c8-11ea-8a40-36850278d945.png)

With this, properties and methods that are only for product editing will still be hidden inside `EditableProductEntity`. Other modules will not be able to use them. 

---

Again, this is just for discussion. 🙂 This may actually be over-engineering! Let's first observe how the app evolves and be vigilant of the effects of extensions.

Reply:
Thanks for the further explanation and diagram! I understand more about the potential issues now. yea, certain extensions methods/variables just don't make sense out of the feature's scope. in this case, I think `var productTaxStatus: ProductTaxStatus` and `var backordersSetting: ProductBackordersSetting` make sense to be accessible within the app layer and can be in an extension of the data models like the common properties protocol. I created a ticket to do this after M3: https://github.com/woocommerce/woocommerce-ios/issues/2575